### DISCUSSION TITLE
Generate Release Notes - Using lookup_a_work_item

### DISCUSSION THREAD

**Discussion by JosephKuper** [rfennell/AzurePipelines#1517] [Q&A] [answered]
Using **Generate Release Notes**, the tool is powerful with lots of capability. The documentation on certain aspects, however, seem to lack or I can't find enough examples of use for these concepts.

When using **lookup_a_work_item**, what I am trying to do is for the pull/build, I want to see the **EPIC** then **ALL workitems** associated with it (children).

With the code snippet below, I can see the epic and get the detail of it. I can see all the associated children but I cannot get the child informaiton (only the **this.url** is showing so I can see the **child IDs** I want). 

**How do I use _lookup_a_work_item_ to get the information of a WorkItem when I know the Item's ID? 
What is this function actually expecting? What are the expected parameter types?**

**Note:** with (lookup_a_work_item ../../../../../relatedWorkItems  this.url), I tried all the tree levels but still can't get that data:

-  ../../../../../relatedWorkItems
-  ../../../../relatedWorkItems
-  ../../../relatedWorkItems
-  ../../relatedWorkItems
- ../relatedWorkItems

**The Code:**
{{#forEach this.relations}}
{{#if (contains this.attributes.name 'Parent')}}
{{#with (lookup_a_work_item ../../relatedWorkItems  this.url)}}
Release for: {{this.id}} - {{lookup this.fields 'System.Title'}} - Type: {{lookup this.fields 'System.WorkItemType'}} 
{{this.url}}    
    
    {{#forEach this.relations}}
    {{#with (lookup_a_work_item ../../../../../relatedWorkItems  this.url)}}   

        {{this.url}}
        - {{this.id}} - {{lookup this.fields 'System.Title'}} - Type: {{lookup this.fields 'System.WorkItemType'}} 
    
    {{/with}}  
    {{/forEach}}

{{/with}}    
{{/if}}
{{/forEach}}

{{/with}}    
{{/if}}
{{/forEach}}

If I remove the "**{{#with (lookup_a_work_item ../../../../../relatedWorkItems  this.url)}**}" from the code, I can see the URLs with the IDs so that means it working mostly:
![image](https://github.com/rfennell/AzurePipelines/assets/135735917/a913d800-1453-43cb-8e2c-5a1fd0d9791e)

**Comment by JosephKuper** [comment]
Sorry, I do not know how to put Labels in a discussion, this was supposed to be with: [GenerateReleaseNotes (XPlat version)]

**Comment by rfennell** [comment, marked as answer]
Sorry you did not find what you needed in the documentation. I do make a serious effort document all the features, but I cannot cover all the edge cases.

A few points that could be of use

## ../ & Tree Depth
The number of `../` tree levels required depends on the nesting of your `foreach` loops. All the arrays of objects are in the root, so if you want to reference an array from within a top level `foreach` it is `../`, If you have two nested `foreach` loops it is `../../` etc. That said it is confusing and there I aways have to use a bit of trial and error.

## Is the data you want present?
You are looking in the `relatedWorkItems` array for these child WI, but are the ones you need present in this array? 

There are two parameters relevant here to populate this array
- `getParentsAndChildren` Get Direct Parent and Children for associated work items, defaults to false
- `getAllParents`  Get All Parents for associated work items, recursing back to workitems with no parents e.g. up to Epics, defaults to false

The first gets **DIRECT** parents and children i.e. one level to an associated WI to a commit
The second starts with a WI associated with a commit and work **UP** the tree until their is no parent

I think neither of these will provide the result you want i.e. all the children of an Epic. Assuming the Epic is associated with the commit the best you can get the the direct children i,e, Features.

## Debugging
To dig into this more have you used the [command line tester](https://github.com/rfennell/AzurePipelines/blob/main/Extensions/XplatGenerateReleaseNotes/XplatGenerateReleaseNotesTask/testconsole/readme.md), this allows easy rerunning and experimentation with templates.

Also I would recommend setting the `DumpPayloadToFile` parameter so you get a JSON representation of the data generated by the task that is passed into the template. Viewing this in a JSON editor can make things clearer.

**Reply by JosephKuper** [reply]
Thanks so much for the response and details. My inability to see what I was looking for through your documentation is likely due to the pressure of the task I was working on. 

I was able to pull the WorkItems using the WIQL feature that your tool lavages, from there I could pull the associated items from the build.

Thanks again for the response and this awesome solution!
